otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py

```diff
@@ -171,3 +191,202 @@ def _balance_losses(self, losses_cls, losses_reg):
         loss_reg = torch.exp(-self.loss_weights[1]) * loss_reg + 0.5 * self.loss_weights[1]
 
         return (loss_cls, loss_reg)
+
+
+@HEADS.register_module()
+class CustomSSDHeadTrackingLossDynamics(CustomSSDHead):
+    """CustomSSDHead which supports tracking loss dynamics."""
+
+    def loss(self, cls_scores, bbox_preds, gt_bboxes, gt_labels, img_metas, gt_bboxes_ignore=None):
+        """Compute loss from the head and prepare for loss dynamics tracking."""
+        self.cur_loss_idx = 0
+        self.loss_dyns = {
+            TrackingLossType.cls: defaultdict(LossAccumulator),
+            TrackingLossType.bbox: defaultdict(LossAccumulator),
+        }
+        return super().loss(cls_scores, bbox_preds, gt_bboxes, gt_labels, img_metas, gt_bboxes_ignore)
+
+    def loss_single(
+        self, cls_score, bbox_pred, anchor, labels, label_weights, bbox_targets, bbox_weights, num_total_samples
+    ):
+        """Compute loss of a single image and increase `self.cur_loss_idx` counter for loss dynamics tracking."""
+        losses = super().loss_single(
+            cls_score, bbox_pred, anchor, labels, label_weights, bbox_targets, bbox_weights, num_total_samples
+        )
+        self.cur_loss_idx += 1
+        return losses
+
+    def _store_loss_dyns(self, losses: torch.Tensor, key: TrackingLossType) -> None:
+        loss_dyns = self.loss_dyns[key]
+        for batch_idx, bbox_idx, loss_item in zip(self.batch_inds, self.bbox_inds, losses.detach().cpu()):
+            loss_dyns[(batch_idx.item(), bbox_idx.item())].add(loss_item.item())
+
+    def _get_pos_inds(self, labels):
+        pos_inds = super()._get_pos_inds(labels)
+
+        if len(pos_inds) > 0:
+            pos_assigned_gt_inds = self.all_pos_assigned_gt_inds[self.cur_loss_idx].reshape(-1)
+
+            gt_inds = pos_assigned_gt_inds[pos_inds].cpu()
+
+            self.batch_inds = gt_inds // self.max_gt_bboxes_len
+            self.bbox_inds = gt_inds % self.max_gt_bboxes_len
+
+        self.pos_inds = pos_inds
+        return pos_inds
+
+    def _get_loss_cls(self, num_total_samples, loss_cls_all, pos_inds, topk_loss_cls_neg):
+        loss_cls_pos = loss_cls_all[pos_inds]
+        loss_cls_neg = topk_loss_cls_neg.sum()
+        loss_cls = (loss_cls_pos.sum() + loss_cls_neg) / num_total_samples
+
+        self._store_loss_dyns(loss_cls_pos.detach(), TrackingLossType.cls)
```

Comment:
If we move this to loss function for `TrackingLossDynamics`, we may be reuse the `CustomSSDHead`'s one too.

Reply:
https://github.com/openvinotoolkit/training_extensions/pull/2115/commits/cbf7ecc05780fc23d5d854a59942beb0d401f539 tries to move the shared codes between ATSS and SSD to `class TrackingLossDynamicsMixIn` as possible:
https://github.com/openvinotoolkit/training_extensions/blob/cbf7ecc05780fc23d5d854a59942beb0d401f539/otx/algorithms/detection/adapters/mmdet/models/heads/cross_dataset_detector_head.py#L283-L289